pkg/vpa/vpa.go

```diff
@@ -247,40 +297,69 @@ func (r Reconciler) createVPA(namespace, vpaName string, updateMode v1beta2.Upda
 	}
 
 	if !r.DryRun {
-		klog.Infof("Creating vpa: %s", vpaName)
-		klog.V(9).Infof("%v", vpa)
+		klog.V(9).Infof("Creating VPA/%s: %v", vpa.Name, vpa)
 		_, err := r.VPAClient.Client.AutoscalingV1beta2().VerticalPodAutoscalers(namespace).Create(vpa)
 		if err != nil {
-			klog.Errorf("Error creating vpa: %v", err)
+			klog.Errorf("Error creating VPA/%s with vpa-update-mode=%s in Namespace/%s: %v", vpaName, updateMode, namespace, err)
 			return err
 		}
+		klog.Infof("Created VPA/%s with vpa-update-mode=%s in Namespace/%s", vpaName, updateMode, namespace)
 	} else {
-		klog.Infof("Dry run was set. Not creating vpa: %v", vpaName)
+		klog.Infof("Not creating VPA/%s with vpa-update-mode=%s in Namespace/%s due to dryrun.", vpaName, updateMode, namespace)
 	}
 	return nil
 }
 
-func vpaUpdateModeForNamespace(ns *corev1.Namespace) v1beta2.UpdateMode {
-	// See: https://github.com/kubernetes/autoscaler/blob/master/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2/types.go#L101
-	updateMode := v1beta2.UpdateModeOff
-	for k, v := range ns.GetLabels() {
-		if strings.ToLower(k) != vpaUpdateModeLabel {
-			continue
+func (r Reconciler) updateVPA(namespace string, vpa *v1beta2.VerticalPodAutoscaler) error {
+	if !r.DryRun {
+		klog.V(9).Infof("Updating VPA/%s: %v", vpa.Name, vpa)
+		_, err := r.VPAClient.Client.AutoscalingV1beta2().VerticalPodAutoscalers(namespace).Update(vpa)
+		if err != nil {
+			klog.Errorf("Error updating VPA/%s with vpa-update-mode=%s in Namespace/%s: %v", vpa.Name, *vpa.Spec.UpdatePolicy.UpdateMode, namespace, err)
+			return err
 		}
-		switch strings.ToLower(v) {
-		case "off":
-			updateMode = v1beta2.UpdateModeOff
-		case "auto":
-			updateMode = v1beta2.UpdateModeAuto
-		case "initial":
-			updateMode = v1beta2.UpdateModeInitial
-		case "recreate":
-			updateMode = v1beta2.UpdateModeRecreate
-		default:
-			klog.Warningf("Found unsupported value for vpaUpdateMode label: %s, using default vpa-update-mode=off", v)
-			updateMode = v1beta2.UpdateModeOff
+		klog.Infof("Updated VPA/%s with vpa-update-mode=%s in Namespace/%s", vpa.Name, *vpa.Spec.UpdatePolicy.UpdateMode, namespace)
+	} else {
+		klog.Infof("Not updating VPA/%s with vpa-update-mode=%s due to dryrun.", vpa.Name, *vpa.Spec.UpdatePolicy.UpdateMode)
+	}
+	return nil
+}
+
+// vpaUpdateModeForResource searches the resource's annotations and labels for a vpa-update-mode
+// key/value and uses that key/value to return the proper UpdateMode type
+func vpaUpdateModeForResource(obj runtime.Object) v1beta2.UpdateMode {
+	var requestedVpaMode string
+
+	// check for vpa-update-mode in annotations first
+	accessor, _ := meta.Accessor(obj)
```

Comment:
I did not know this was a thing! Very cool. This will help me in another project.